Projects/Checkers/Engine.cs

```diff
@@ -0,0 +1,274 @@
+﻿namespace Checkers;
+
+public static class Engine
+{
+	public static MoveOutcome? PlayNextMove(PieceColor side, Board board, (int X, int Y)? from = null, (int X, int Y)? to = null)
+	{
+		List<Move> possibleMoves;
+		MoveOutcome? outcome;
+		if (from is not null && to is not null)
+		{
+			outcome = GetAllPossiblePlayerMoves(side, board, out possibleMoves);
+			if (possibleMoves.Count is 0)
+			{
+				outcome = MoveOutcome.NoMoveAvailable;
+			}
+			else
+			{
+				if (MoveIsValid(from, to.Value, possibleMoves, board, out Move? selectedMove))
+				{
+					possibleMoves.Clear();
+					if (selectedMove is not null)
+					{
+						possibleMoves.Add(selectedMove);
+						outcome = selectedMove.Capturing is not null ? MoveOutcome.Capture : MoveOutcome.ValidMoves;
+					}
+				}
+			}
+		}
+		else
+		{
+			outcome = AnalysePosition(side, board, out possibleMoves);
+		}
+		// If a side can't play then other side wins
+		if (outcome == MoveOutcome.NoMoveAvailable)
+		{
+			outcome = side == PieceColor.Black ? MoveOutcome.WhiteWin : MoveOutcome.BlackWin;
+		}
+		switch (outcome)
+		{
+			case MoveOutcome.EndGame:
+			case MoveOutcome.ValidMoves:
+				Move bestMove = possibleMoves[Random.Shared.Next(possibleMoves.Count)];
+				Piece pieceToMove = bestMove.PieceToMove!;
+				int newX = bestMove.To.X;
+				int newY = bestMove.To.Y;
+				pieceToMove.XPosition = newX;
+				pieceToMove.YPosition = newY;
+				// Promotion can only happen if not already a king and you have reached the far side
+				if (newY is 0 or 7 && !pieceToMove.Promoted)
+				{
+					pieceToMove.Promoted = true;
+				}
+				break;
+			case MoveOutcome.Capture:
+				bool anyPromoted = PerformCapture(side, possibleMoves, board);
+				bool moreAvailable = MoreCapturesAvailable(side, board);
+				if (moreAvailable && !anyPromoted)
+				{
+					outcome = MoveOutcome.CaptureMoreAvailable;
+				}
+				break;
+		}
+		if (outcome is not MoveOutcome.CaptureMoreAvailable)
+		{
+			board.Aggressor = null;
```

Comment:
There is a bug here. When a human player has an "Aggressor" (meaning they can perform additional captures), if they attempt to move another piece that is not the aggressor, the aggressor is still being reset. So it allows a human player to skip on multiple captures in favor of moving another piece, which is not valid in the rules of Checkers.

Reply:
this bug was fixed in b4675fc917a579c078048b96c714b0d8dd3f0243